### PR TITLE
Param ids fix

### LINKED ISSUE #1085: pytest raises an invalid UnicodeEncodeError in Python 2.7
This looks like it has just crept into pytest 2.8.1.

I have a parametrised test where I'm passing in unicode, decoded from bytes. Pytest fails to even collect the test. This didn't used to happen (env's with 2.7.2 are running fine) and I'm pretty sure the code is legitimate across Python 2/3

```
class TestASCII:

    def test_valid(self, ascii):
        ascii.value = b'some text'
        assert ascii.value == b'some text'

    value = b'\xc3\xbc'.decode("utf-8")
    @pytest.mark.parametrize("value",
                             [
                                 value,
                                 10,
                                 []
                             ]
                             )
    def test_invalid(self, ascii, value):
        with pytest.raises(TypeError):
            ascii.value = value
```

### LINKED ISSUE #1087: SystemError when passing an empty byte string in Python 3.x
Under Python 3.x with pytest 2.8.1

```python
import pytest

@pytest.mark.parametrize("message", [b""])
def test_broken(message):
    pass
```

causes a SystemError during the call to `codecs.escape_encode` inside `_escape_bytes`. This was introduced in commit https://github.com/pytest-dev/pytest/commit/e1063678f10b48c5872f6e9d829c21be9097748b, but is really a CPython bug as `codecs.escape_encode(b'')` should not cause a SystemError. That said, pytest probably needs to work around the bug for now.

Stack trace:
```
../../.virtualenvs/crypto3/lib/python3.4/site-packages/_pytest/runner.py:149: in __init__
    self.result = func()
../../.virtualenvs/crypto3/lib/python3.4/site-packages/_pytest/main.py:437: in _memocollect
    return self._memoizedcall('_collected', lambda: list(self.collect()))
../../.virtualenvs/crypto3/lib/python3.4/site-packages/_pytest/main.py:317: in _memoizedcall
    res = function()
../../.virtualenvs/crypto3/lib/python3.4/site-packages/_pytest/main.py:437: in <lambda>
    return self._memoizedcall('_collected', lambda: list(self.collect()))
../../.virtualenvs/crypto3/lib/python3.4/site-packages/_pytest/python.py:578: in collect
    return super(Module, self).collect()
../../.virtualenvs/crypto3/lib/python3.4/site-packages/_pytest/python.py:432: in collect
    res = self.makeitem(name, obj)
../../.virtualenvs/crypto3/lib/python3.4/site-packages/_pytest/python.py:444: in makeitem
    collector=self, name=name, obj=obj)
../../.virtualenvs/crypto3/lib/python3.4/site-packages/_pytest/vendored_packages/pluggy.py:724: in __call__
    return self._hookexec(self, self._nonwrappers + self._wrappers, kwargs)
../../.virtualenvs/crypto3/lib/python3.4/site-packages/_pytest/vendored_packages/pluggy.py:338: in _hookexec
    return self._inner_hookexec(hook, methods, kwargs)
../../.virtualenvs/crypto3/lib/python3.4/site-packages/_pytest/vendored_packages/pluggy.py:333: in <lambda>
    _MultiCall(methods, kwargs, hook.spec_opts).execute()
../../.virtualenvs/crypto3/lib/python3.4/site-packages/_pytest/vendored_packages/pluggy.py:595: in execute
    return _wrapped_call(hook_impl.function(*args), self.execute)
../../.virtualenvs/crypto3/lib/python3.4/site-packages/_pytest/vendored_packages/pluggy.py:249: in _wrapped_call
    wrap_controller.send(call_outcome)
../../.virtualenvs/crypto3/lib/python3.4/site-packages/_pytest/python.py:307: in pytest_pycollect_makeitem
    res = list(collector._genfunctions(name, obj))
../../.virtualenvs/crypto3/lib/python3.4/site-packages/_pytest/python.py:464: in _genfunctions
    self.ihook.pytest_generate_tests(metafunc=metafunc)
../../.virtualenvs/crypto3/lib/python3.4/site-packages/_pytest/vendored_packages/pluggy.py:724: in __call__
    return self._hookexec(self, self._nonwrappers + self._wrappers, kwargs)
../../.virtualenvs/crypto3/lib/python3.4/site-packages/_pytest/vendored_packages/pluggy.py:338: in _hookexec
    return self._inner_hookexec(hook, methods, kwargs)
../../.virtualenvs/crypto3/lib/python3.4/site-packages/_pytest/vendored_packages/pluggy.py:333: in <lambda>
    _MultiCall(methods, kwargs, hook.spec_opts).execute()
../../.virtualenvs/crypto3/lib/python3.4/site-packages/_pytest/vendored_packages/pluggy.py:596: in execute
    res = hook_impl.function(*args)
../../.virtualenvs/crypto3/lib/python3.4/site-packages/_pytest/python.py:217: in pytest_generate_tests
    metafunc.parametrize(*marker.args, **marker.kwargs)
../../.virtualenvs/crypto3/lib/python3.4/site-packages/_pytest/python.py:994: in parametrize
    ids = idmaker(argnames, argvalues, idfn)
../../.virtualenvs/crypto3/lib/python3.4/site-packages/_pytest/python.py:1108: in idmaker
    for valindex, valset in enumerate(argvalues)]
../../.virtualenvs/crypto3/lib/python3.4/site-packages/_pytest/python.py:1108: in <listcomp>
    for valindex, valset in enumerate(argvalues)]
../../.virtualenvs/crypto3/lib/python3.4/site-packages/_pytest/python.py:1103: in _idvalset
    for val, argname in zip(valset, argnames)]
../../.virtualenvs/crypto3/lib/python3.4/site-packages/_pytest/python.py:1103: in <listcomp>
    for val, argname in zip(valset, argnames)]
../../.virtualenvs/crypto3/lib/python3.4/site-packages/_pytest/python.py:1082: in _idval
    return _escape_bytes(val)
../../.virtualenvs/crypto3/lib/python3.4/site-packages/_pytest/python.py:1058: in _escape_bytes
    encoded_bytes, _ = codecs.escape_encode(val)
E   SystemError: Objects/bytesobject.c:2828: bad argument to internal function
```

### PR DESCRIPTION
**Pull request by nicoddemus**
Fix #1087 
 Fix #1085

### CODE REVIEW

File: _pytest/python.py [hunk 1]
```diff
@@ -1064,7 +1068,7 @@ def _escape_bytes(val):
         is a full ascii string, otherwise escape it into its binary form.
         """
         try:
-            return val.encode('ascii')
+            return val.decode('ascii')
         except UnicodeDecodeError:
             return val.encode('string-escape')
```
Comment (RonnyPfannschmidt) [inline _pytest/python.py]: im not 100% sure off-hand is string-escape correctly used that way?

Comment (nicoddemus) [inline _pytest/python.py]: It seems so:

```
Python 2.7.10 (default, May 23 2015, 09:40:32) [MSC v.1500 32 bit (Intel)] on win32
Type "help", "copyright", "credits" or "license" for more information.
>>> u'ação'
u'a\xe7\xe3o'
>>> u'ação'.encode('utf-8')
'a\xc3\xa7\xc3\xa3o'
>>> s=_
>>> s.encode('string-escape')
'a\\xc3\\xa7\\xc3\\xa3o'
```

File: _pytest/python.py [hunk 2]
```diff
@@ -1053,10 +1053,14 @@ def _escape_bytes(val):
            want to return escaped bytes for any byte, even if they match
            a utf-8 string.
         """
-        # source: http://goo.gl/bGsnwC
-        import codecs
-        encoded_bytes, _ = codecs.escape_encode(val)
-        return encoded_bytes.decode('ascii')
+        if val:
+            # source: http://goo.gl/bGsnwC
+            import codecs
```
Comment (hpk42) [inline _pytest/python.py]: i guess it makes sense to move "import codecs" outside the function (within the py3 if section)

Comment (nicoddemus) [inline _pytest/python.py]: Agreed, thanks!

Comment (nicoddemus) [inline _pytest/python.py]: Done!

**Comment by nicoddemus** [comment]
Anything else guys?

**Comment by RonnyPfannschmidt** [comment]
FML, now it has conflicts, 
The travisci failure was a hickup of travis

**Comment by nicoddemus** [comment]
Fixed merge conflict. :sweat_smile:

**Comment by RonnyPfannschmidt** [comment]
:+1:

**Pull request merged**